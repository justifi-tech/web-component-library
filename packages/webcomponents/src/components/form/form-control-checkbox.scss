@import 'reboot';
@import 'forms';

:host {
  @include host-base-styles;
  display: block;
}

// input[type='checkbox'][disabled] {
//   color: var(--jfi-form-control-disabled-color);
//   opacity: 0.5;
//   cursor: not-allowed;
// }

*::part(input[type='checkbox']) {
  margin: var(--jfi-form-control-margin);
  padding: var(--jfi-form-control-padding);
}