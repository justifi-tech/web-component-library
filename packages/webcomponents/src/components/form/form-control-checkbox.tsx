import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';

@Component({
  tag: 'form-control-checkbox'
})
export class CheckboxInput {
  checkboxElement!: HTMLInputElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue?: boolean;
  @Prop() inputHandler: (name: string, value: boolean) => void;
  @Prop() disabled: boolean;
  
  @Watch('defaultValue')
  handleDefaultValueChange(newValue: boolean) {
    this.updateInput(newValue);
  }

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.checked);
    this.formControlInput.emit({ name, value: target.value });
  }

  updateInput = (newValue: any) => {
    this.checkboxElement.checked = newValue;
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <div class='form-group d-flex flex-column'>
          <div class="form-check">
            <input
              ref={el => (this.checkboxElement = el as HTMLInputElement)}
              type="checkbox"
              id={this.name}
              name={this.name}
              onBlur={this.formControlBlur.emit}
              onInput={this.handleFormControlInput}
              part={`input-checkbox ${this.errorText ? 'input-checkbox-invalid' : ''}`}
              class={this.errorText ? 'form-check-input is-invalid' : 'form-check-input'}
              disabled={this.disabled}
            />
            <label class="form-check-label" htmlFor={this.name}>
              {this.label}
            </label>
          </div>
          <form-control-help-text helpText={this.helpText} name={this.name} />
          <form-control-error-text errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
