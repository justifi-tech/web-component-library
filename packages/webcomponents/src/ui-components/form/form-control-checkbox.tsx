import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
  State,
} from '@stencil/core';
import { FormControlHelpText, FormControlErrorText } from '../../ui-components';
import { inputCheckbox, inputCheckboxChecked, inputCheckboxCheckedFocused, inputCheckboxFocused, inputCheckboxInvalid, label } from '../../styles/parts';

@Component({
  tag: 'form-control-checkbox',
})
export class CheckboxInput {
  checkboxElement!: HTMLInputElement;

  @State() isFocused: boolean = false;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue?: boolean;
  @Prop() inputHandler: (name: string, value: boolean) => void;
  @Prop() disabled: boolean;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: boolean) {
    this.updateInput(newValue);
  }

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.checked);
    this.formControlInput.emit({ name, value: target.value });
  }

  updateInput = (newValue: any) => {
    this.checkboxElement.checked = newValue;
  }

  private get part() {
    if (this.errorText) {
      return inputCheckboxInvalid;
    }
    if (this.isFocused && this.checkboxElement.checked) {
      return inputCheckboxCheckedFocused;
    }
    if (this.isFocused) {
      return inputCheckboxFocused;
    }
    if (this.checkboxElement.checked) {
      return inputCheckboxChecked;
    }
    return inputCheckbox;
  }

  render() {
    return (
      <Host>
        <div class='form-group d-flex flex-column'>
          <div class="form-check">
            <input
              ref={el => (this.checkboxElement = el as HTMLInputElement)}
              type="checkbox"
              id={this.name}
              name={this.name}
              onFocus={() => this.isFocused = true}
              onBlur={() => {
                this.isFocused = false;
                this.formControlBlur.emit();
              }}
              onInput={this.handleFormControlInput}
              part={this.part}
              class={this.errorText ? 'form-check-input is-invalid' : 'form-check-input'}
              disabled={this.disabled}
            />
            <label class="form-check-label" htmlFor={this.name} part={label}>
              {this.label}
            </label>
          </div>
          <FormControlHelpText helpText={this.helpText} name={this.name} />
          <FormControlErrorText errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
