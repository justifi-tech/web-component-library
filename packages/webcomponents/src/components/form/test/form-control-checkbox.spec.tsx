import { newSpecPage } from '@stencil/core/testing';
import { CheckboxInput } from '../form-control-checkbox';

describe('form-control-checkbox', () => {

  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [CheckboxInput],
      html: `<form-control-checkbox label="Select a checkbox" name="user ID"></form-control-checkbox>`,
    });

    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.label).toBe('Select a checkbox');
    expect(page.rootInstance.name).toBe('user ID');
  });

  it('renders with all props provided', async () => {
    const page = await newSpecPage({
      components: [CheckboxInput],
      html: `
        <form-control-checkbox
          label="Select a checkbox"
          name="email"
          error="No checkbox selected"
          disabled
        ></form-control-checkbox>
      `,
    });

    const inputElement = page.root.shadowRoot.querySelector('input');
    expect(page.rootInstance.label).toBe('Select a checkbox');
    expect(inputElement.disabled).toBeTruthy();
  });

  it('handles user input correctly', async () => {
    const page = await newSpecPage({
      components: [CheckboxInput],
      html: `<form-control-checkbox></form-control-checkbox>`,
    });

    const inputElement = page.root.shadowRoot.querySelector('input');
    const testValue = true;

    inputElement.checked = testValue;
    await inputElement.dispatchEvent(new Event('input'));

    expect(inputElement.checked).toBe(testValue);
  });

  it('emits formControlInput event on input', async () => {
    const page = await newSpecPage({
      components: [CheckboxInput],
      html: `<form-control-checkbox></form-control-checkbox>`,
    });

    // Set a mock inputHandler to prevent it from being undefined
    page.rootInstance.inputHandler = jest.fn();

    const inputEventSpy = jest.fn();
    page.root.addEventListener('formControlInput', inputEventSpy);
  });
});
