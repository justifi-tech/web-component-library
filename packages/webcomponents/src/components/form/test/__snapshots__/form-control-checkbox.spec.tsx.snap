// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-checkbox renders with default props 1`] = `
<form-control-checkbox exportparts="label,input,input-invalid" label="Select a checkbox" name="user ID">
  <mock:shadow-root>
    <label class="form-label" htmlfor="user ID" part="label">
      Select a checkbox
    </label>
    <input class="form-select" id="user ID" name="user ID" part="input " type="checkbox">
  </mock:shadow-root>
</form-control-checkbox>
`;
